File: src/server/routes/estoque.ts
import { Router } from 'express';
import sqlite3 from 'sqlite3';

const router = Router();
const db = new sqlite3.Database('akay-sales.db');

// Cria a tabela de estoque se não existir
// Estrutura baseada em ProdutoPrimario
// id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem

db.run(`
  CREATE TABLE IF NOT EXISTS estoque (
    id TEXT PRIMARY KEY,
    produto TEXT,
    peso REAL,
    medida TEXT,
    preco REAL,
    quantidade INTEGER,
    quantidadeAtual INTEGER,
    periodo TEXT,
    vencimento TEXT,
    imagem TEXT,
    tipo TEXT CHECK(tipo IN ('acabamentos','corantes','embalagens','essencias','primario','propriedades'))
  )
`);

// Listar todos os itens do estoque, com filtro opcional por tipo
router.get('/estoque', (req, res) => {
    const { tipo } = req.query;
    let sql = 'SELECT * FROM estoque';
    const params: any[] = [];
    if (tipo) {
        sql += ' WHERE tipo = ?';
        params.push(tipo);
    }
    db.all(sql, params, (err, rows) => {
        if (err) return res.status(500).json({ error: err.message });
        res.json(rows);
    });
});

// Criar novo item de estoque
router.post('/estoque', (req, res) => {
    const { id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo } = req.body;
    console.log("Dados recebidos para criar estoque:", req.body);
    // {"id": 147, "produto": "Válvula Gatilho Branca G28", "peso": 0, "medida": "28", "preco": 2.1, "quantidade": 2, "quantidadeAtual": 2, "periodo": "17/05/2025", "vencimento": "", "imagem": "", "tipo": "acabamentos"}
    db.run(
        'INSERT INTO estoque (id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)',
        [id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo],
        function (err) {
            if (err) {
                console.error("Erro ao inserir no estoque:", err.message);
                return res.status(500).json({ error: err.message });
            }
            try {
                res.json({ id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo });
            } catch (error) {
                res.status(500).json({ error: 'Erro ao criar item de estoque' });
            }
        });
});

// Atualizar item de estoque
router.put('/estoque/:id', (req, res) => {
    const { produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo } = req.body;
    db.run(
        'UPDATE estoque SET produto = ?, peso = ?, medida = ?, preco = ?, quantidade = ?, quantidadeAtual = ?, periodo = ?, vencimento = ?, imagem = ?, tipo = ? WHERE id = ?',
        [produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo, req.params.id],
        function (err) {
            if (err) return res.status(500).json({ error: err.message });
            res.json({ id: req.params.id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo });
        }
    );
});


// Deletar item de estoque
router.delete('/estoque/:id', (req, res) => {
    db.run('DELETE FROM estoque WHERE id = ?', [req.params.id], function (err) {
        if (err) return res.status(500).json({ error: err.message });
        res.json({ success: true });
    });
});

// Drop na tabela estoque
router.post('/estoque/drop', (req, res) => {
    db.run('DROP TABLE IF EXISTS estoque', function (err) {
        if (err) return res.status(500).json({ error: err.message });
        res.json({ success: true, message: 'Tabela estoque removida.' });
    });
});

export default router;

File: src/sql/estoque.py
import sqlite3
import requests
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
db_path = PROJECT_ROOT / 'server' / 'akay-sales.db'

def migrate_quantidade_to_real(db_path=db_path):
    """
    Altera as colunas quantidade e quantidadeAtual da tabela estoque para REAL (float).
    Cria uma tabela temporária, copia os dados, remove a antiga e renomeia a nova.
    """
    conn = sqlite3.connect(db_path)
    cur = conn.cursor()
    try:
        cur.execute('''
            CREATE TABLE IF NOT EXISTS estoque_temp (
                id TEXT PRIMARY KEY,
                produto TEXT,
                peso REAL,
                medida TEXT,
                preco REAL,
                quantidade REAL,
                quantidadeAtual REAL,
                periodo TEXT,
                vencimento TEXT,
                imagem TEXT,
                tipo TEXT
            )
        ''')
        cur.execute('''
            INSERT INTO estoque_temp (id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo)
            SELECT id, produto, peso, medida, preco, quantidade, quantidadeAtual, periodo, vencimento, imagem, tipo FROM estoque
        ''')
        cur.execute('DROP TABLE estoque')
        cur.execute('ALTER TABLE estoque_temp RENAME TO estoque')
        conn.commit()
        print('Migração concluída: quantidade e quantidadeAtual agora são REAL.')
    except Exception as e:
        print('Erro na migração:', e)
        conn.rollback()
    finally:
        conn.close()

API_URL = 'http://localhost:4000/api/estoque'

def delete_all_items():
    resp = requests.get(API_URL)
    if resp.status_code == 200:
        for item in resp.json():
            del_resp = requests.delete(f'{API_URL}/{item["id"]}')
            print(f'Deletado: {item["produto"]} | Status: {del_resp.status_code}')
    else:
        print('Erro ao buscar estoque:', resp.text)

def drop_estoque_table():
    # Requer rota específica no backend para dropar a tabela
    resp = requests.post(f'{API_URL}/drop')
    print('Drop tabela estoque | Status:', resp.status_code, '|', resp.text)

def update_item(item_id, data):
    resp = requests.put(f'{API_URL}/{item_id}', json=data)
    print(f'Atualizado: {item_id} | Status: {resp.status_code} | {resp.text}')

def get_all_items():
    resp = requests.get(API_URL)
    if resp.status_code == 200:
        return resp.json()
    else:
        print('Erro ao buscar estoque:', resp.text)
        return []

def get_item(item_id):
    resp = requests.get(f'{API_URL}/{item_id}')
    if resp.status_code == 200:
        return resp.json()
    else:
        print('Erro ao buscar item:', resp.text)
        return None

if __name__ == "__main__":
    # Exemplo de uso: deletar todos os itens
    migrate_quantidade_to_real()

